torch/nn/attention/_flex_attention.py

```diff
@@ -39,11 +39,127 @@ def _identity(
+        B, H, Q // Q_BLOCK_SIZE, Q_BLOCK_SIZE, KV // KV_BLOCK_SIZE, KV_BLOCK_SIZE
+    )  # [B, H, Q//Q_BLOCK_SIZE, Q_BLOCK_SIZE, KV//KV_BLOCK_SIZE, KV_BLOCK_SIZE]
+    mask = mask.permute(
+        0, 1, 2, 4, 3, 5
+    )  # [B, H, Q//Q_BLOCK_SIZE, KV//KV_BLOCK_SIZE, Q_BLOCK_SIZE, KV_BLOCK_SIZE]
+    mask = mask.sum(dim=[-2, -1]) > 0  # [B, H, Q//Q_BLOCK_SIZE, KV//KV_BLOCK_SIZE]
+    return mask
+
+
+def _convert_block_mask_to_mask(
+    block_mask,
+    KV_BLOCK_SIZE=_DEFAULT_SPARSE_BLOCK_SIZE,
+    Q_BLOCK_SIZE=_DEFAULT_SPARSE_BLOCK_SIZE,
+):
+    assert block_mask.dim() == 4
+    B, H, Q, KV = block_mask.shape
+    block_mask = block_mask.expand(Q_BLOCK_SIZE, KV_BLOCK_SIZE, *block_mask.shape)
+    block_mask = block_mask.permute(2, 3, 4, 0, 5, 1).reshape(
+        B, H, Q * Q_BLOCK_SIZE, KV * KV_BLOCK_SIZE
+    )
+    return block_mask
+
+
+def _create_block_sparse_mask(
+    mask: torch.Tensor,
+    KV_BLOCK_SIZE: int = _DEFAULT_SPARSE_BLOCK_SIZE,
+    Q_BLOCK_SIZE: int = _DEFAULT_SPARSE_BLOCK_SIZE,
+):
+    block_mask = _convert_mask_to_block_mask(
+        mask, KV_BLOCK_SIZE=KV_BLOCK_SIZE, Q_BLOCK_SIZE=Q_BLOCK_SIZE
+    )
+    block_mask = block_mask.to(dtype=torch.int8)
+    kv_num_blocks = block_mask.sum(dim=3)
+    kv_indices = torch.argsort(block_mask, dim=3, descending=True, stable=True)
+    q_num_blocks = block_mask.sum(dim=2)
+    q_indices = torch.argsort(block_mask, dim=2, descending=True, stable=True).permute(
+        0, 1, 3, 2
+    )
+    return _BlockSparseMask(
+        kv_num_blocks=kv_num_blocks.to(torch.int32).to(mask.device).contiguous(),
+        kv_indices=kv_indices.to(torch.int32).to(mask.device).contiguous(),
+        q_num_blocks=q_num_blocks.to(torch.int32).to(mask.device).contiguous(),
+        q_indices=q_indices.to(torch.int32).to(mask.device).contiguous(),
+        KV_BLOCK_SIZE=KV_BLOCK_SIZE,
+        Q_BLOCK_SIZE=Q_BLOCK_SIZE,
+    )
+
+
+"""
+    The flex attention kernels are implemented using block sparsity,
+    where only the unmasked blocks are computed to get the best perf.
+    If users don't specify any block sparse mask info, we create this
+    empty block sparse mask with all blocks unmasked as the default one.
+"""
+
+
+def _create_empty_block_sparse_mask(query, key, value):
+    device = query.device
+    kv_len = key.size()[-2]
+    q_len = query.size()[-2]
+    return _BlockSparseMask(
+        kv_num_blocks=torch.ones([1, 1, 1], dtype=torch.int32, device=device),
+        kv_indices=torch.zeros([1, 1, 1, 1], dtype=torch.int32, device=device),
+        q_num_blocks=torch.ones([1, 1, 1], dtype=torch.int32, device=device),
+        q_indices=torch.zeros([1, 1, 1, 1], dtype=torch.int32, device=device),
+        KV_BLOCK_SIZE=kv_len,
+        Q_BLOCK_SIZE=q_len,
+    )
+
+
 def _flex_attention(
     query: torch.Tensor,
     key: torch.Tensor,
     value: torch.Tensor,
     score_mod: _score_mod_signature = _identity,
+    block_sparse_mask: Optional[_BlockSparseMask] = None,
```

Comment:
I think we should just call this `mask` btw.

Reply:
Will update in the follow-ups.